### PR TITLE
Address thread safety for lambda processor and lambda sink

### PR DESCRIPTION
**Pull request by srikanthjg**
### Description

Following changes are addressed in this commit:
Lambda Processor:
1. Address thread safety
2. Fix metrics

Lambda Sink:
1. Release events gracefully
2. Fix metrics
 
### Issues Resolved
Addresses #4700 
 
### Check List
- [X] New functionality includes testing.
- [] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 1]
```diff
@@ -149,9 +139,25 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
             return records;
         }
 
-        //lambda mutates event
+        // Initialize here to void multi-threading issues
+        // Note: By default, one instance of processor is created across threads.
+        BufferFactory bufferFactory = new InMemoryBufferFactory();
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:144]: i think then it becomes a shared resource among the threads. I wanted to keep all these limited to the scope of the method for less complications.

File: data-prepper-plugins/aws-lambda/src/test/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessorTest.java [hunk 2]
```diff
@@ -1,45 +1,519 @@
+//package org.opensearch.dataprepper.plugins.lambda.processor;
```
Comment (dlvenable) [inline data-prepper-plugins/aws-lambda/src/test/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessorTest.java:1]: Please remove all of this commented out code.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 3]
```diff
@@ -72,26 +73,26 @@ public class LambdaProcessor extends AbstractProcessor<Record<Event>, Record<Eve
     private final String invocationType;
     private final List<String> tagsOnMatchFailure;
     private final BatchOptions batchOptions;
-    private final BufferFactory bufferFactory;
     private final LambdaAsyncClient lambdaAsyncClient;
     private final AtomicLong requestPayloadMetric;
     private final AtomicLong responsePayloadMetric;
     OutputCodecContext codecContext = null;
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:79]: yes as a part of the json creation, this is needed

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 4]
```diff
@@ -149,9 +139,25 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
...
+        BufferFactory bufferFactory = new InMemoryBufferFactory();
+        lambdaCommonHandler = new LambdaCommonHandler(LOG, lambdaAsyncClient, functionName, invocationType);
+        Buffer currentBufferPerBatch = lambdaCommonHandler.createBuffer(bufferFactory);
+        List futureList = new ArrayList<>();
+        totalFlushedEvents = 0;
+
+        // Setup request codec
+        JsonOutputCodecConfig jsonOutputCodecConfig = new JsonOutputCodecConfig();
+        jsonOutputCodecConfig.setKeyName(batchOptions.getKeyName());
+        OutputCodec requestCodec = new JsonOutputCodec(jsonOutputCodecConfig);
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:153]: The input codec can be created once.

Looking at the code, the output codec does not look like it supports multiple threads.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 5]
```diff
@@ -163,24 +169,33 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
...
+                if (wasFlushed) {
+                    currentBufferPerBatch = lambdaCommonHandler.createBuffer(bufferFactory);
+                    requestCodec = new JsonOutputCodec(jsonOutputCodecConfig);
+                }
             } catch (Exception e) {
                 LOG.error(NOISY, "Exception while processing event {}", event, e);
                 handleFailure(e, currentBufferPerBatch, resultRecords);
-                currentBufferPerBatch.reset();
+                currentBufferPerBatch = lambdaCommonHandler.createBuffer(bufferFactory);
+                requestCodec = new JsonOutputCodec(jsonOutputCodecConfig);
```
Comment (dlvenable) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:189]: Why are we doing this in a catch?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java (hunk 5 above)
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:189]: this resets the buffer and codec for the "remaining events" ; the next call to flushToLambdaIfNeeded needs it.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:153]: Yes, OutputCodec is not thread safe specifically for jsonOutputCodec:     JsonGenerator generator and OutputCodecContext codecContext are mutable fields. 
Regarding inputCodec, i guess it is ok to keep it thread safe in the event something changes with inputCodec iteself?

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 6]
```diff
@@ -163,24 +169,33 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
...
                 currentBufferPerBatch.addRecord(record);
 
-                flushToLambdaIfNeeded(resultRecords, false);
+                boolean wasFlushed =  flushToLambdaIfNeeded(resultRecords, currentBufferPerBatch,
+                        requestCodec, responseCodec,futureList,false);
+
+                // After flushing, create a new buffer for the next batch
+                if (wasFlushed) {
+                    currentBufferPerBatch = lambdaCommonHandler.createBuffer(bufferFactory);
+                    requestCodec = new JsonOutputCodec(jsonOutputCodecConfig);
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:183]: These have to be reset for every flush and it has to be in the loop.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:141]: InMemoryBufferFactory can definitely be in constructor. Please see it's implementation. It has no members

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 7]
```diff
@@ -149,9 +136,25 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
...
+        // Initialize here to void multi-threading issues
+        // Note: By default, one instance of processor is created across threads.
+        BufferFactory bufferFactory = new InMemoryBufferFactory();
+        lambdaCommonHandler = new LambdaCommonHandler(LOG, lambdaAsyncClient, functionName, invocationType);
+        Buffer currentBufferPerBatch = lambdaCommonHandler.createBuffer(bufferFactory);
+        List futureList = new ArrayList<>();
+        totalFlushedEvents = 0;
+
+        // Setup request codec
+        JsonOutputCodecConfig jsonOutputCodecConfig = new JsonOutputCodecConfig();
```
Comment (kkondaka) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:148]: Can this config be moved to constructor?

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java (hunk 5 above)
Comment (dlvenable) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:186]: Should this be in a `finally` instead?

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java [hunk 8]
```diff
@@ -189,52 +201,59 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
...
                 futureList.add(processingFuture);
-
-                // Create a new buffer for the next batch
-                currentBufferPerBatch = lambdaCommonHandler.createBuffer(currentBufferPerBatch);
             } catch (IOException e) {
                 LOG.error(NOISY, "Exception while flushing to lambda", e);
                 handleFailure(e, currentBufferPerBatch, resultRecords);
-                currentBufferPerBatch = lambdaCommonHandler.createBuffer(currentBufferPerBatch);
+            } finally {
+                return true;
```
Comment (dlvenable) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:249]: I'm not sure it makes sense to have a `return` in a `finally` block. I think you should just have `return true` after the block.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java (hunk 7 above)
Comment (srikanthjg) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java:148]: might need additional changes to tests, will work on it on a follow up PR.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/AggregateResponseEventHandlingStrategy.java [hunk 9]
```diff
@@ -5,13 +5,19 @@
...
 import java.util.List;
 
 public class AggregateResponseEventHandlingStrategy implements ResponseEventHandlingStrategy {
 
+    private static final Logger LOG = LoggerFactory.getLogger(AggregateResponseEventHandlingStrategy.class);
+
     @Override
-    public void handleEvents(List<Event> parsedEvents, List<Record<Event>> originalRecords, List<Record<Event>> resultRecords, Buffer flushedBuffer) {
+    public void handleEvents(List<Event> parsedEvents, List<Record<Event>> originalRecords,
+                             List<Record<Event>> resultRecords, Buffer flushedBuffer) {
```
Comment (sb2k16) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/AggregateResponseEventHandlingStrategy.java:19]: do we need to pass the `flushedBuffer` as a parameter? we are not using it here.

**Review by sb2k16 — APPROVED** [review]

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/sink/LambdaSinkService.java [hunk 10]
```diff
@@ -167,6 +169,12 @@ public void output(Collection<Record<Event>> records) {
         // Wait for all futures to complete
         lambdaCommonHandler.waitForFutures(futureList);
 
+        // Release event handles for records not sent to Lambda
```
Comment (kkondaka) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/sink/LambdaSinkService.java:172]: Please remove this in the next PR. There is no "conditional" in sink.

Comment (kkondaka) [reply data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/sink/LambdaSinkService.java:172]: This actually results in duplicate release but it is a no-op.

File: data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/sink/LambdaSinkService.java [hunk 11]
```diff
@@ -107,14 +108,13 @@ public LambdaSinkService(final LambdaAsyncClient lambdaAsyncClient, final Lambda
...
-        events = new ArrayList();
-        futureList = new ArrayList<>();
+        futureList = Collections.synchronizedList(new ArrayList<>());
 
         this.bufferFactory = bufferFactory;
 
         LOG.info("LambdaFunctionName:{} , invocationType:{}", functionName, invocationType);
         // Initialize LambdaCommonHandler
-        lambdaCommonHandler = new LambdaCommonHandler(LOG, lambdaAsyncClient, functionName, invocationType, bufferFactory);
+        lambdaCommonHandler = new LambdaCommonHandler(LOG, lambdaAsyncClient, functionName, invocationType);
```
Comment (sb2k16) [inline data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/sink/LambdaSinkService.java:117]: do we need to pass the `LOG`.

**Review by kkondaka — APPROVED** [review]

**Pull request merged**